util/types/json/functions.go

```diff
@@ -165,3 +165,117 @@ func extract(j JSON, pathExpr PathExpression) (ret []JSON) {
 	}
 	return
 }
+
+// autoWrapAsArray wraps input JSON into an array if needed.
+func autoWrapAsArray(j JSON, hintLength int) JSON {
+	jnew := CreateJSON(nil)
+	jnew.typeCode = typeCodeArray
+	jnew.array = make([]JSON, 0, hintLength)
+	jnew.array = append(jnew.array, j)
+	return jnew
+}
+
+// Merge merges suffixes into j according the following rules:
+// 1) adjacent arrays are merged to a single array;
+// 2) adjacent object are merged to a single object;
+// 3) a scalar value is autowrapped as an array before merge;
+// 4) an adjacent array and object are merged by autowrapping the object as an array.
+func (j JSON) Merge(suffixes []JSON) JSON {
+	if j.typeCode != typeCodeArray && j.typeCode != typeCodeObject {
+		j = autoWrapAsArray(j, len(suffixes)+1)
+	}
+	for i := 0; i < len(suffixes); i++ {
+		suffix := suffixes[i]
+		switch j.typeCode {
+		case typeCodeArray:
+			if suffix.typeCode == typeCodeArray {
+				// rule (1)
+				j.array = append(j.array, suffix.array...)
+			} else {
+				// rule (3), (4)
+				j.array = append(j.array, suffix)
+			}
+		case typeCodeObject:
+			if suffix.typeCode == typeCodeObject {
+				// rule (2)
+				for key := range suffix.object {
+					j.object[key] = suffix.object[key]
+				}
+			} else {
+				// rule (4), notice here we should retry that suffix again.
+				j = autoWrapAsArray(j, len(suffixes)+1-i)
+				i--
```

Comment:
it's true that modify i inside the loop is not clear, so I commented that.
but, we can not simply `j.array = append(j.array, suffix.array)` because
we don't know `suffix.typeCode == typeCodeArray`.